File: Python/288. Unique Word Abbreviation.py
class ValidWordAbbr:

    def __init__(self, dictionary: List[str]):
        self.abbr = collections.defaultdict(int)
        self.dictionary = set(dictionary)
        for word in self.dictionary:
            if len(word) <= 2:
                self.abbr[word] += 1
            else:
                self.abbr[word[0] + str(len(word[1:-1])) + word[-1]] += 1

    def isUnique(self, word: str) -> bool:        
        if len(word) <= 2:
            abr = word
        else:
            abr = word[0] + str(len(word[1:-1])) + word[-1]        
        return abr not in self.abbr or (self.abbr[abr] == 1 and word in self.dictionary)
        


# Your ValidWordAbbr object will be instantiated and called as such:
# obj = ValidWordAbbr(dictionary)
# param_1 = obj.isUnique(word)

File: 2021/288. Unique Word Abbreviation.cpp
class ValidWordAbbr {
public:
    unordered_map<string, int> abbr;
    vector<string> dict;
    ValidWordAbbr(vector<string>& dictionary) {        
        dict = dictionary;
        for (int i = 0; i < dictionary.size(); i++) {
            string& word = dictionary[i];
            string ab = word.substr(0, 1) + ((word.length() > 2) ? to_string(word.length() - 2) : "") + word.substr(word.length() -  1, 1);                                    
            if (abbr.find(ab) != abbr.end() && abbr[ab] != -1 && dict[abbr[ab]] != word) {                
                abbr[ab] = -1;
            } else {
                abbr[ab] = i;
            }            
        }
    }
    
    bool isUnique(string word) {
         string ab = word.substr(0, 1) + ((word.length() > 2) ? to_string(word.length() - 2) : "") + word.substr(word.length() - 1, 1);
        return abbr.find(ab) == abbr.end() || (abbr[ab] != -1 && dict[abbr[ab]] == word);                             
    }
};

/**
 * Your ValidWordAbbr object will be instantiated and called as such:
 * ValidWordAbbr* obj = new ValidWordAbbr(dictionary);
 * bool param_1 = obj->isUnique(word);
 */